File: day12/sort_algorithm/06_quick_sort.py
class Solution:
    def quick(self, nums, l, r):
        if l >= r:
            return nums
        first = l
        end = r
        while l < r:
            while l < r and nums[r] > nums[first]:
                r -= 1
            while l < r and nums[l] <= nums[first]:
                l += 1
            
            nums[l], nums[r] = nums[r], nums[l]
        nums[l], nums[first] = nums[first], nums[l]
        self.quick(nums, first, l - 1)
        self.quick(nums, l + 1, end)
        return nums

    def sortArray(self, nums: List[int]) -> List[int]:
        return self.quick(nums, 0, len(nums) - 1)

File: day12/sort_algorithm/06_QuickSort.java
class Solution {

    private static void swap(int nums[], int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
        return;
    }

    public static void quickSort(int nums[], int left, int right){
        if (left >= right) { return; }
        int pivot = left;
        int end = right;
        while (left < right) {
            while (left < right && nums[right] >= nums[pivot]) {
                right--;
            }
            while (left < right && nums[left] <= nums[pivot]) {
                left++;
            }
            swap(nums, left, right);
        }
        swap(nums, left, pivot);
        quickSort(nums, pivot, left - 1);
        quickSort(nums, left + 1, end);
    }
        
    public int[] sortArray(int[] nums) {
        quickSort(nums, 0, nums.length - 1);
        return nums;
    }
}
